cma/evolution_strategy.py

```diff
@@ -4646,4 +4661,30 @@ def update(es):
     es.objective_function_complements = [_al]
     es.augmented_lagrangian = _al
     es.best_feasible = best_feasible_solution
+
+    if post_optimization:
+        positive_constraints = np.where(np.array(g(es.result.xfavorite)) > 0)
+        if len(positive_constraints[0]) > 0:
+            x_post_opt, es_post_opt = fmin_con(lambda x: sum([gi**2 if gi > 0 else 0 for gi in g(x)]),
```

Comment:
Shouldn't we terminate the optimization the moment a feasible solution is found? This should be possible by setting the `'ftarget'` option to value `0`.

Reply:
I tried to implement this by adding the following code:
`kwargs_post_opt = kwargs.copy()`
 `if 'options' in kwargs_post_opt:`
 `     kwargs_post_opt['options']['ftarget'] = 0`
 ` else:`
 `     kwargs_post_opt['optins'] = {'ftarget': 0}`

However, I notice that it rarely results in a feasible solution (see graph bellow)

![Figure_1](https://user-images.githubusercontent.com/63016186/144468373-e239bd94-fe1c-4474-99dd-51357fd03d3f.png)

I understand that taking the first feasible solution is usefull to gain some time but the solution given by `es_post_opt.best_feasible` is always feasible and at least as good as the first feasible solution found. But I don't know if in practise, saving some computation time is worth the loss in accuracy.

.

Reply:
If this _is_ the code you tested: the last line has a typo in the keyword.

AFAICS it is not possible that the best solution after the target was reached is not feasible.

Reply:
For posting code, use three upticks and add "python" after the first, like

```
``python
some_code
``
```
but with three instead of two ticks.

Reply:
Right, the test actually went through the line `kwargs_post_opt['options']['ftarget'] = 0`. So I still have the same issue. I agree that with `ftarget=0`, the best solution should be feasible but when I test it, I get positive constraints (of the order 1e-4). Here is the updated code:

```
python 
            kwargs_post_opt = kwargs.copy()
            if 'options' in kwargs_post_opt:
                kwargs_post_opt['options']['ftarget'] = 0
            else:
                kwargs_post_opt['options'] = {'ftarget': 0}
            x_post_opt, es_post_opt = fmin_con(lambda x: sum([gi**2 if gi > 0 else 0 for gi in g(x)]),
                                               es.result.xfavorite, sigma0 / 1000, g=g, h=h, **kwargs_post_opt)
```

I checked and in my test, `kwargs_post_opt['options']` is the following:

`{'seed': 1, 'verbose': -9, 'tolstagnation': 0, 'ftarget': 0}`

Reply:
Right, good point, the `ftarget` trick can only work reliably if the post optimization does not use the constraints. Using `max(g(es.mean))` as callback is not a workable option for two reasons:

- `es.mean` is not a solution which is supposed to be evaluated, because it is a "genotype"
- it costs an additional evaluation which should be avoided if possible

Not passing the inequality constraints seems fine, but the question is how to deal with the equality constraints, also because it is not possible to get a feasible solution then.

Reply:
Ok, if I understand correctly, the idea would be to use `fmin2()` rather than `fmin_con()` to avoid the use of the Augmented Lagrangian and to be able to use the `ftarget=0` termination. But the problem is that we won't take the equality constraint into account in that case.

I don't know how to answer this problem.

Reply:
Right, makes sense, but the latter depends on `eps`. One could require the `eps` from the user (in case we have equality constraints), like, in case when equality constraints are present and a "feasible" solution is requested, then `post_optimization` must be a strictly positive `float` indicating the error on the inequality constraints.

Reply:
Ok, I modified the code implement this. It is included in the last commit.